chctools/horn_hacking_lemmas.py

```diff
@@ -0,0 +1,710 @@
+    PInitForm0 = pInit(*(newParList + [0]))
+    PInitForm1 = pInit(*(newParList + [1]))
+
+    for r in db._rules:
+        r = replace_func(r, newParList)
+        if is_rule_pre(r):
+            if len(newParList) > 0:
+                swapped = swap_const_U(r, copy.deepcopy(newParList[0]))
+                if not swapped:
+                    print("Could not swap in U.")
+                    return
+            
+            r._rels.append(pInit)
+            r._body.append(PInitForm1)
+            r._formula = r.mk_formula()
+
+    secondRule = HornRule(z3.ForAll(newParList, \
+        z3.Implies(PInitForm0, PInitForm1)))
+    secondRule._update()
+    secondRule.mk_formula()
+    db.add_rule(secondRule)
+
+    firstRule = HornRule(z3.ForAll(newParList, \
+    z3.Implies(pInitPre, PInitForm0)))
+    firstRule._update()
+    firstRule.mk_formula()
+    db.add_rule(firstRule)
+
+    db.seal()
+    return db
+
+
+def swap_const_U(rule, param):
+    #flips first seen constant to U
+    body = rule._body
+    def is_num_val(val):
+        if z3.is_const(val) \
+            and (val.decl().kind() == z3.Z3_OP_ANUM or val.decl().kind() == z3.Z3_OP_AGNUM): #i.e. is fixed value!
+            #Otherwise: "and val.decl().kind() != z3.Z3_OP_UNINTERPRETED:""
+                return True
+        if val.decl().kind() == z3.Z3_OP_TO_REAL or val.decl().kind() == z3.Z3_OP_TO_INT:
+            return True
+        return False
+
+    for i in range(len(body)): #TODO: Remove Assumption: HACKY SPECIAL CASES #reversed to swap diff
+        e = body[i]
+        if z3.is_expr(e) and e.decl().kind() == z3.Z3_OP_EQ:
+            #for j in range(e.num_args()):
+            val0 = param if is_num_val(e.arg(0)) else e.arg(0)
+            val1 = param if is_num_val(e.arg(1)) else e.arg(1)
+            body[i] = val0 == val1
+            return True #success
+        elif z3.is_expr(e) and e.decl().kind() == z3.Z3_OP_LE:
+            val0 = param if is_num_val(e.arg(0)) else e.arg(0)
+            val1 = param if is_num_val(e.arg(1)) else e.arg(1)
+            body[i] = val0 <= val1
+            return True
+        elif z3.is_expr(e) and e.decl().kind() == z3.Z3_OP_GE:
+            val0 = param if is_num_val(e.arg(0)) else e.arg(0)
+            val1 = param if is_num_val(e.arg(1)) else e.arg(1)
+            body[i] = val0 >= val1
+            return True
+        elif z3.is_expr(e) and e.decl().kind() == z3.Z3_OP_LT:
+            val0 = param if is_num_val(e.arg(0)) else e.arg(0)
+            val1 = param if is_num_val(e.arg(1)) else e.arg(1)
+            body[i] = val0 < val1
+            return True
+        elif z3.is_expr(e) and e.decl().kind() == z3.Z3_OP_GT:
+            val0 = param if is_num_val(e.arg(0)) else e.arg(0)
+            val1 = param if is_num_val(e.arg(1)) else e.arg(1)
+            body[i] = val0 > val1
+            return True
+    return False
+
+
+def is_rule_pre(rule):
```

Comment:
not sure what this checks for.

Reply:
Checking for the (initial rule in original benchmark) of form: "init(x) => Inv(x)" by inspecting structure.

I mutate them into "PInit(u) ^init(x) => Inv(x, u)" rule of new benchmarks.